File: src/infra/token_repository/in_memory.rs
use std::sync::Mutex;

use super::{DeleteError, InsertError, TokenRepository};
use crate::domain::models::{Id, Token};
use async_trait::async_trait;

pub struct InMemoryTokenRepository {
    error: bool,
    tokens: Mutex<Vec<(Id, Token)>>,
}

impl InMemoryTokenRepository {
    #[cfg(test)]
    pub fn new() -> Self {
        Self {
            error: false,
            tokens: Mutex::new(vec![]),
        }
    }

    #[cfg(test)]
    pub fn with_error(self) -> Self {
        Self {
            error: true,
            ..self
        }
    }
}

#[async_trait]
impl TokenRepository for InMemoryTokenRepository {
    async fn delete(&self, token: Token) -> Result<Id, DeleteError> {
        if self.error {
            return Err(DeleteError::Unknown("error flag is on"));
        }

        let mut lock = match self.tokens.lock() {
            Ok(lock) => lock,
            _ => return Err(DeleteError::Unknown("could not acquire lock")),
        };

        let index = match lock.iter().position(|(_, t)| t == &token) {
            Some(index) => index,
            None => return Err(DeleteError::Unknown("token not found")),
        };

        let (salary_id, _) = lock.remove(index);

        Ok(salary_id)
    }

    async fn insert(&self, salary_id: Id, token: Token) -> Result<(), InsertError> {
        if self.error {
            return Err(InsertError::Unknown("error flag is on"));
        }

        let mut lock = match self.tokens.lock() {
            Ok(lock) => lock,
            _ => return Err(InsertError::Unknown("could not acquire lock")),
        };

        lock.push((salary_id, token));

        Ok(())
    }
}

File: src/infra/tokenRepository/inMemory.ts
import { InMemoryRepository } from "../utils/mod.ts";
import { TokenRepository } from "./mod.ts";
import { Id, Token } from "@domain";
import { Result } from "@utils";

export class InMemoryTokenRepository implements TokenRepository {
    private repo: InMemoryRepository<{ id: Id, token: Token }>;

    private constructor(tokens: { id: Id, token: Token }[], error: boolean) {
        this.repo = new InMemoryRepository(tokens, error);
    }

    static new() {
        return new InMemoryTokenRepository([], false);
    }

    static withError() {
        return new InMemoryTokenRepository([], true);
    }

    async delete(token: Token): Promise<Result<Id, string>> {
        const deleteResult = await this.repo.delete({ filter: t => t.token.raw === token.raw });
        
        return Result.map(
            deleteResult,
            ({ id }) => id
        );
    }

    insert(id: Id, token: Token): Promise<Result<void, string>> {
        return this.repo.insert({ id, token });
    }
}
